### DISCUSSION TITLE
V8: Reusable components in TypeScript (using branded strings)

### DISCUSSION THREAD

**Discussion by felixschorer** [react-hook-form/react-hook-form#7354] [RFC]
# The Problem

Consider an app which has a model like this: An `Address` interface which is used by both the `Person` and `Order` interfaces.
```ts
interface Address {
  street: string
  houseNumber: number
  city: string
  country: string
}

interface Person {
  name: string
  address: Address
}

interface Order {
  item: string
  shippingAddress: Address
}
```

In Javascript we can easily create a generic `AddressFields` component for editing the address which can be either embedded in the `PersonFields` or the `OrderFields` component.

```jsx
function AddressFields({ register, path }) {
  return <div>
    <input type="text" {...register(`${path}.street`)} />
    <input type="number" {...register(`${path}.houseNumber`)} />
    <input type="text" {...register(`${path}.city`)} />
    <input type="text" {...register(`${path}.country`)} />
  </div>
}

function PersonFields({ register }) {
  return <div>
    <input type="text" {...register('name')} />
    <AddressFields register={register} path="address" />
  </div>
}

function OrderFields({ register }) {
  return <div>
    <input type="text" {...register('item')} />
    <AddressFields register={register} path="shippingAddress" />
  </div>
}
```

In Typescript this is currently almost impossible (almost because it could be done with overloads or type assertions).

# Branded strings to the rescue

In Typescript you can associate extra type information with a type using type assertions: [Branded types](https://medium.com/@KevinBGreene/surviving-the-typescript-ecosystem-branding-and-type-tagging-6cf6e516523d).

We could use this technique to associate a path with the type of the form values and its resolved type. E.g.
```ts
type PathWithMeta<FormValues, Type> = string & { __formValues: FormValues, __type: Type }
```

We can then define basic functions for creating and joining such branded strings.
```ts
function of<FormValues, Key extends keyof FormValues>(key: Key) {
  return key as PathWithMeta<FieldValues, FieldValues[Key]>
}

function join<FieldValues, Type, Key extends keyof Type>(
  path: PathWithMeta<FieldValues, Type>, 
  key: Key,
) {
  return `${path}.${key}` as PathWithMeta<FieldValues, Type[Key]>
}
```

That would then allow us to use the same pattern which we've used in Javascript.
```tsx
function AddressFields<FieldValues>(props: {
  register: UseFormRegister<FieldValues>, 
  path: PathWithMeta<FieldValues, Address>,
}) {
  const { register, path } = props
  return <div>
    <input type="text" {...register(join(path, 'street')} />
    <input type="number" {...register(join(path, 'houseNumber'))} />
    <input type="text" {...register(join(path, 'city'))} />
    <input type="text" {...register(join(path, 'country'))} />
  </div>
}

function PersonFields(props: { register: UseFormRegister<Person> }) {
  const { register } = props
  return <div>
    <input type="text" {...register('name')} />
    <AddressFields 
      register={register} 
      path={of<Person, "address">('address')} 
    />
  </div>
}

function OrderFields(props: { register: UseFormRegister<Order> }) {
  const { register } = props
  return <div>
    <input type="text" {...register('item')} />
    <AddressFields 
      register={register}
      path={of<Order, "shippingAddress">('shippingAddress')}
    />
  </div>
}
```

# Required Changes

- The changes are mostly to the types. Every function which currently accepts path strings would need an overload to also accept `PathWithMeta`.
- Optionally, the functionality of `of` and `join` could be integrated into the functions directly.
  ```ts
  register(path, 'street')
  // or
  const registerAddress: UseFormRegister<Address> = register.for('address')
  registerAddress('street')
  ```
- `useForm` could also return a bound version of `of` which would make the type annotations obsolete.

# Related issues

https://github.com/react-hook-form/react-hook-form/issues/5055
https://github.com/react-hook-form/react-hook-form/discussions/7413

cc @kotarella1110 @jorisre @barrymay

**Comment by bluebill1049** [comment]
I love this RFC and this is a proper solution to address generic component type difficulty with `of` and `join`.  

> Optionally, the functionality of of and join could be integrated into the functions directly

How could we avoid breaking change? Ideally, I hope we can avoid changing/add `register` API, as this simple API is growing a little bit too fast over the years.

**Reply by felixschorer** [reply]
I am not a fan of that either. I am just listing some alternatives.

**Comment by felixschorer** [comment]
Actually, once all functions are converted to using `LazyPath` the path argument would be constrained to just `string`. `PathWithMeta` **is** a `string`. So we wouldn't even have to touch the overloads. It would just be a change to the types which work on paths. I.e. just `PathValue`.

So the entire feature would just be two functions, a new type `PathWithMeta`, and a conditional around `PathValue`.

```ts
type PathValue<T, P extends string> = 
  P extends PathWithMeta<T, infer U>
    ? U
    : PathValueImpl<T, P>;
```

**Reply by bluebill1049** [reply]
yea, that's it. I think by moveing`FieldPath` into `string` will make the generic component typing easier.

**Reply by felixschorer** [reply]
Yes, definitely.

**Comment by barrymay** [comment]
Would this be able to fit into the useFormContext model? Sounds like it would be a great fit.

**Reply by felixschorer** [reply]
I don't think so. The context API can't be well typed once the context is allowed to be generic. You loose all information about the type arguments. You will always have to annotate the type arguments of `useFormContext` which is basically an implicit type assertion. So I would personally avoid that feature in TS and rather pass `register` down the component tree through props.

**Reply by bluebill1049** [reply]
Good suggestion, but yea we will need to pass down `register` (uncontrolled) component and `control` (controlled) component.

**Reply by felixschorer** [reply]
Or just pass just pass the entire `UseFormReturn` down 😉

**Reply by felixschorer** [reply]
I mean the purpose of this feature is to enable strict type safety. That may require a bit more boilerplate code, but I am willing to pay that price.

**Reply by bluebill1049** [reply]
yep, 💯

**Reply by felixschorer** [reply]
To clarify, you can absolutely use it like that.

```tsx
function AddressFields<FieldValues>(props: {
  path: PathWithMeta<FieldValues, Address>,
}) {
  const { path } = props
  const { register } = useFormConext<FieldValues>()
  return <div>
    <input type="text" {...register(join(path, 'street')} />
    <input type="number" {...register(join(path, 'houseNumber'))} />
    <input type="text" {...register(join(path, 'city'))} />
    <input type="text" {...register(join(path, 'country'))} />
  </div>
}
```

However, since `useFormConext<FieldValues>()` is basically `useFormContext() as UseFormReturn<FieldValues>`, you don't gain much type safety.

**Reply by barrymay** [reply]
Agreed that for fully strict type safety it may not work but with a more advanced custom hook there may be some options where you don't need to pass down the path prop, yet you can still get that useful register back from the hook. (The whole point of context for me is to avoid repetitive prop passing for complex scenarios)

**Comment by felixschorer** [comment]
The type arguments of `of` don't have to be specified, if one annotates the generic of `AddressFields`

```tsx
function PersonFields(props: { register: UseFormRegister<Person> }) {
  const { register } = props
  return <div>
    <input type="text" {...register('name')} />
    <AddressFields<Person> 
      register={register} 
      path={of('address')} 
    />
  </div>
}

function OrderFields(props: { register: UseFormRegister<Order> }) {
  const { register } = props
  return <div>
    <input type="text" {...register('item')} />
    <AddressFields<Order> 
      register={register}
      path={of('shippingAddress')}
    />
  </div>
}
```

That even works when using `LazyPath` for the definition of `of`.

**Reply by bluebill1049** [reply]
nice!

**Reply by chrisb2244** [reply]
Is it possible to write a component that doesn't have to be nested? (Is there a 'root' path?)
I have something like your example, and it works well (using 8.0.0-alpha4), but if I want to use the <AddressFields> without nesting it in <PersonFields> or <OrderFields> I can't find a way to pass `path`.

In my case, I have a Person I want to reuse, and it looks something like:
```
import ...
import { TextField, ... } from '@mui/material'

export type PersonProps = {
  firstName: string
  lastName: string
  email: string
}

export function Person<FieldValues>(props: {
  register: UseFormRegister<FieldValues>
  path: TypedFieldPath<FieldValues, PersonProps>
  defaultValue?: PersonProps
}) {
  const { register, path, defaultValue } = props
  return (
  <>
    <TextField {...register(join(path, 'firstName')), { required: true })} defaultValue={defaultValue?.firstName} />
    <TextField {...register(join(path, 'lastName')), { required: true })} defaultValue={defaultValue?.lastName} />
    <TextField {...register(join(path, 'email')), { required: true })} defaultValue={defaultValue?.email} />
  </>
  )
}
```

and when I use it like this it works fine:

```
type BiggerFormData = {
  firstPerson: PersonProps
  otherPeople: PersonProps[]
  ... other stuff ... 
}

// I couldn't make it work without an intermediate component to handle the array elements
const OtherPerson = <FieldValues,>(props: { 
  idx: number
  path: TypedFieldPath<FieldValues, PersonProps[]>
  register: UseFormRegister<FieldValues>
}) => {
  return <Person path={join(props.path, `${props.idx}`)} register={props.register} />
}

export const Blah = ({...}) => {
  const { register, control } = useForm<BiggerFormData>()
  const { fields, append, remove } = useFieldArray<BiggerFormData, 'otherPeople'>({ name: 'otherPeople', control })
  return (
    <>
      <Person<BiggerFormData> register={register} path={of('firstPerson')} />
       {fields?.map((p, idx) => {
         return <>
              <OtherPerson
                idx={idx}
                path={of<BiggerFormData, 'otherPeople', PersonProps[]>(
                  'otherPeople'
                )}
                register={register}
              />
              <Button onClick={() => remove(idx)}>Remove</Button>
            </>
          )
         <Button onClick={() => append({})}>Add Person</Button>
    </>
  )
}
```

But I can't find a way to not pass `path` to `Person`, like in
```
const { control, handleSubmit, reset, register } = useForm<PersonProps>({
    mode: 'onBlur'
  })
// What to put for path?
<Person<PersonProps> register={register} path={of('')} />
```

Undefined and null, or attempting to conditionally join ran into type errors every attempt (although probably I missed something obvious...)

One workaround would presumably be to in that component use
```
type WrappedPerson = { person: PersonProps }
...
const { register } = useForm<WrappedPerson >(...)
<Person<WrappedPerson > register={register} path={of('person')} />
```
but that seems a bit undesirable if it isn't required.

**Reply by felixschorer** [reply]
Thanks for bringing that up. That's something which I haven't considered yet. 
There needs to be an empty path primitive. 

You should be able to work around this like that:
```tsx
function EMPTY<T>() {
  return '' as TypedFieldPath<T, T>
}

<Person<PersonProps> register={register} path={EMPTY()} />
```

**Reply by chrisb2244** [reply]
Thanks, that's a nice workaround! I'm still wrapping my head around the names for the generic types here.

**Comment by matamatanot** [comment]
ref #4739

**Reply by bluebill1049** [reply]
hey @matamatanot thanks for the input. Notice in the example 

```tsx
// which you still need to extends FormValues
const Input = <TFormValues extends IFormValues1 | IFormValues2>({
``` 

This new proposal is more flexible in terms of `infer` nested form. 

If you refer to the following example, this `AddressFields` now can be placed inside any nested component. 

```tsx
function AddressFields<FieldValues>(props: {
  register: UseFormRegister<FieldValues>, 
  path: PathWithMeta<FieldValues, Address>,
}) {
  const { register, path } = props
  return <div>
    <input type="text" {...register(join(path, 'street')} />
    <input type="number" {...register(join(path, 'houseNumber'))} />
    <input type="text" {...register(join(path, 'city'))} />
    <input type="text" {...register(join(path, 'country'))} />
  </div>
}

<AddressFields path="bill" />
<AddressFields path="matamatanot" />
```

**Comment by arutkowski00** [comment]
Is it also going to work with `control` and `useController` ? I'd like to parametrize the `convertDetails` part of the name while keeping it type-safe, so the section can be reused for different fields with the same model.

```tsx
export interface ConvertSectionProps {
  form: UseFormReturn<DealDetailsType>;
}

export function ConvertSection({
  form: { control, watch },
}: ConvertSectionProps) {
  const currency = watch("currencyCode");

  return (
    <ContentBlock>
      <AutoFormGrid>
        <InputField
          control={control}
          name="convertDetails.amount"
          label={`Amount (${currency})`}
        />
        <InputField
          control={control}
          name="convertDetails.maturityYears"
          label="Maturity (in years)"
        />
      </AutoFormGrid>
    </ContentBlock>
  );
}
```

Great proposal btw, can't wait for that to happen!

**Reply by felixschorer** [reply]
Yes, absolutely. You can give it a try already. There is an alpha version which implements #7389.

https://codesandbox.io/s/falling-cookies-s1xzb?file=/src/App.tsx

There are still some rough edges regarding `NestedValue` and `UnpackNestedValue` which requires type assertions for now. That should be finex in the final version.

**Comment by panrafal** [comment]
With some small help from `lodash` and `type-fest` I was able to solve it for v7 with this:

https://gist.github.com/panrafal/92e11be4401aee710d8616cf2b664c0d

It's fully type-safe and rather easy to use. I guess it should handle every bit of RHF as it's mostly some type magic.

```ts
type NestedValues = {
  foo: string
}
type Values = {
  nested: NestedValues
}

function FieldsToBeNested({form}: {form: NestedForm<NestedValues>> {
  const errors = form.get(form.formState.errors)
  return <div><input {...register(form.path('foo'))}/>{errors.foo?.message}</div>
}

function DirectForm() {
  const form = useForm<NestedValues>()
  return <FieldsToBeNested form={nestedForm(form)} />
}

function ComposedForm() {
  const form = useForm<Values>()
  return <FieldsToBeNested form={nestedForm(form, 'nested')} />
}

```

**Reply by jdpt0** [reply]
@panrafal This is great, thanks! Is it possible to get it working for optional nested form fragments? I'm currently getting errors when trying to pass a key that is potentially undefined. 
![image](https://user-images.githubusercontent.com/19761394/181134228-7c91305d-a934-4c94-bf4e-2c9b42a16d72.png)

**Reply by BrendanC23** [reply]
This looks really useful. Does anyone know how to modify it to handle TTransformedValues (a form schema that uses Zod transforms, and so has different input and output types?)

https://github.com/react-hook-form/react-hook-form/blob/ed23a09f75b0f095277ff77aa884820910628058/src/useForm.ts#L49

**Reply by BrendanC23** [reply]
I updated the types to be 

```
export function nestedForm(
    form: UseFormReturn<any, any, any> | NestedForm<any>,
    path?: string | number,
): NestedForm<any>
```

which fixed the issue with `TTransformedValues`.

One other issue that I haven't been able to solve is how to handle a schema where the form has more props than the nested form needs. Using the example from the first post:

```

interface Address {
  street: string
  houseNumber: number
  city: string
  country: string
}

interface PersonAddress {
  street: string
  houseNumber: number
  city: string
  country: string
  personName: string
}

interface Person {
  name: string
  address: PersonAddress 
}
```

Here, `Person` has an extra `personName` field that isn't needed by the nested address component. This causes a type error if I use `NestedForm`.

**Reply by BrendanC23** [reply]
This solution also has an issue with `FieldArray`. It uses [type-fest's Get type](https://github.com/sindresorhus/type-fest/blob/main/source/get.d.ts), which includes `undefined` in the return type by default. This can be fixed by changing the return type of `nestedForm` to

```ts
NestedForm<Get<TValues, TPath, { strict: false }>>;
```

As indicated by [this issue](https://github.com/sindresorhus/type-fest/issues/567), type-fest's `Get` also doesn't support a `number` type in the path (it returns `unknown`). So something like this doesn't work to handle forms with an array:

```ts
const index: number = 0;
const nested = nestedForm(form, `array.${index}.values`);
```

For now, the best workaround is

```ts
const index: number = 0;
const nested = nestedForm(form, `array.${index as 0}.values`);
```

**Comment by jdpt0** [comment]
Is there a way to get this working for more than a depth of one, either in V8 or V7? The solution provided by @panrafal works great for a nested form of one depth but doesn't work for a nested form within another nested form.

**Comment by cbejensen** [comment]
I don't love my current solution, but maybe it will spark some ideas? Just throwing it here in case it's helpful.

https://codesandbox.io/s/react-hook-form-watch-v7-ts-forked-b84nri?file=/src/PersonForm.tsx

**Reply by cbejensen** [reply]
We've started using this home-grown solution with `useNestedForm` instead: https://codesandbox.io/s/use-nested-form-v2-y13f33?file=/src/App.tsx

**Reply by dolma** [reply]
@cbejensen That's a neat solution, thanks! I've tried it and it does indeed solve the problem, however I'm running into some TS error in the `getErrors` function (in `useNestedForm`) and I'm not too sure how to fix it. I'll keep looking on my end but I wonder if you've encountered the same error (or similar) in your own project? And if so, how did you fix it?

![TS Error](https://github.com/react-hook-form/react-hook-form/assets/2726179/1bc2734b-a308-4507-aa79-565e3c0c727f)

**Reply by cbejensen** [reply]
It’s been awhile since I’ve even looked at this code. Don’t supposed you have a code sandbox or something I can look at?

**Reply by dolma** [reply]
Thanks for the quick reply! Sure yeah, here's a link to a TS playgroud with a repro: [TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoUSWOAbzgGEIA7GHAG3aSgBo4AFFDAAWfAGLAk7ACYA1FOwCuSAM59FKpGOggmrJAA8Y4yTICiUHFBVwAvnGy4CRdDAC0wiBADWbzDrJKcGh4egBzJHh7Rzx8dghpFBVhQKoQujgNJAYFdgAjdG87BxxYlxIKchgATzAkOABVZm9mCAB3ZkEROABeAiaW9uY4MCFhODcRriT6rLgImC7hAAoASkDyAHoAKm3yOG24AFkUb1UHHTg0XEhmJFYbIg0UPK4AOn29g7hhGBgwFQALk2m2u0lUKGY0jyEAMb2AEE2Kk2WTcdxUMCQ0j8OjcADcAExuaoARgAzJgyWSAPyYYBcHpIqBoTYAQTAYDeMBUBk+nzgAAFRlAUHhhfdFmMJnARPVRt0YjLhPUcBB4BBMEr6v4oHgYBAlcAbOjMdILrqPgd+QLDKKwFx+QADZ3c3nfNI0ejgzAoRTsGCs6TSIgqFTaXXyJSqPiB4OqMOXaKlAhvNlBkMJ3WBd3BT1wb2+-0AOVFWh0keUajgJZAZd1xUV+FTNbrIGzX0O12YGPzSB9fsW3BULHDIAr5z6tH5cDgzFLgN7-eLpdH47U07gKHT8YXBYHsYzq4Ulf5tgoVu+mEUzAwCOG-CHI50y3oOpASzsC6LqlNo-4OABVY6A3Lse3CSILCsKsFg-ew+iyb8MSxUcAB4ajqDVF0LQdrCfCNj1UAA+ZY3yWVZz0OGdQPgOda0g6AbD6BZ6OsZZ8FopB8HIuAQTgVAVHpap5kiPjLAYvlvhnIgYEUKBhmWDcZxQukpGkTQYEIxSZzgFCuAiKFCIfXDmBQzY9PuaRNMk7SlJbUdzXfMYeloGCxjYjiuPsTYrMomy6A4liVFsLSlIPeN7NIpyXMiJY2K3ONQ08nifL80yVJkdSUpncjTydZ1Pk2chDGoLBr1vFhMk0RDfx0FDqqxNc+KMCybAkVS12I8gAEheIaTQ4EdDEoGAZgwjgAAfTIoT7EasUdOARqQrc4Ew6kZQNEBTnqFRZO23BImEEaxqNEYkk0M0Rs3GxgBgfAbH4oT9XNBZuvFVglgXIajomqbvVm6RyCAqcuuohzekq1s9ExIwUOQa4oGkFCBlaDolj4eq5AIlRCOInLutB1zungzQck4Ao0G8BSuq6lC-ylQxMShGwljqn8GqxwifuvP67jNeDprpXnOupkHDpkJZqQXOmRG6rrVilnQPwZlrfpm3m4DW5Ghg-BdHQAEloLXUbGWw3gNosWCLP12BeLg0NqJBMLQMXpCWQjbHmnpOeB6neIAcXubhgDQNxlahb7rzvGw0EhVp4Dyeo5zEtosWlX5-iBEEwhu4RFDyN5rhATYjmDnBh0wGBNgAFQdgBlNBhrASujR21RNnJABODuAFZZZ6zY+JUdgRvcaQjVtpA0QZtxh7uQV0NUBvgCb0Oh5HzZWlDgx7WDm63Ehao++k2Thmd+lXalNaACsIBGpYVGWN7JREPgz-FsYgIXZY34v7okk3Zg1QcrU1sDwWWABtJ+SwAC63U8YgxYD2ZiYlrDgyyKTfIhQqY02luMMOakBBjFZkhTGUZsZcwFv9NBlChbYNFufCWCtdRLFlvLOAuCmqMwIdzNWqc1ptXMCglQxDTQdTgAuAOMAABC1QWYCOkIFER7MyGEXEIrMYnMvbARFsfOSwkYAkR0G8N8tcYBCCQG8bgUE+CExWD-MiwCuqgIgW+YxOhTHmMsUImxMUxiwLlpULqujhj0DeGEt8PiYCBUibBCgwUti7H5DkYYCdOEWVTnkIS1QICyRGABGw3pLQdkFLacADpvjOkdNyfkxV0gj24D6NA9QjLDmYH+fJaSmaNCqmzaQ7SIAAmAtZGAN0uCfTYEdCidgJKHCvDeEZFUWl4RAC+ByaMZSjPqLYBcSy2mK3yUDECiCsCXGJkgDGo5DHMI-lM4KFT8oHEKrUmg9SoCNPqBc-ZAybDAx2IcKuypToKmTLKOAqp1SalBW+dahpjS9Icm8RgkI4CpNwDdU0C1IWAp-mDLsbAIDsBsMkHJMhNwclQFATc8BQUqFLHALgeIpBXS1KdIgrBEXbEKl1eUwhJZwC+qNOJ5AEnFL6qnJ6N9LpPzBjymwbQc5XBdjKsYKginWlKfaJAeUqkqH5JKzoKq4pvHwHwJs+cTUEDJKa6QxruK8WCTYfAKA3j5zJG8aQ+BtUFSKtvdIczyrDH1ffZYYS3iyq-gKsak1mCKBAAnSlk0eGC1TtG62qxwHQLYZG7RwSgXCF1dTN4+q2K2p4gPeKcBEUJxgCne4fF0DjB5bLN4RB7ToCQMsTYAAdN4tACQ8FsJsMIprS32qQG2pplbEXyu6B8QtrabZNM7QAPR7ZsU1XEy1gqQCACAjL6WoDHqNStzaF3ts7T2vW66CBcVIFuogu791sBQPSb6Hx4l-P5GKs0T0Fibl7OO9gQkTTiodnAPEwAUD-rADgCD4IzTZp5ZVb60g1SzjVEIO8arvg2gMHacphxKnVO+KDCAeQr7g3oHUYyC56AcXGcNUafAUARAXDGuN3A7Bnn5AvOArJwaSJkSzMjV9TXUdafgQid7eJ0fnPyiZgrNysdnLG+N0zvi8akQJyIQmiEibE4+ZgbwPJSa3ZGnjYGGDaekbIvT5GDPGTeCxzipneLsfjRZuocAAAi1ndMiBQvpgg4mWDwmYHiBQwAPWuYHkm-6XrvjpwBMCTY4I8RckRCMgE1QAD6aXNhbgizeLEbgF4qCXivIgI1GWsCOjPBISRhBuAWG4AALJgZU3rnnUrA4J2zAWADy9nCEiE0SNvBzUun61oCNTAnGAAyfYYCmwNrNzjCBgBhF+B7bqa1FsV06QQs42TNRDavrLNafWWZmAMGgJQ4IUJnfAftmA0D1A0KxKoxAm3fic0mjd59GBHvkee0tt7qtk2WVlguOLvNuo7PppNo7SATtwDOxdtHIOYHQ4h-9SoIIeiE6J8TknpOyfk4p5Tqn1Oaek5BFsTYtOmfM5Z6zln9OCds659znntP6eUH0G89tcBa551HPIn53Udp5HStIEkDHJlS7zrLgkbHVPcCFeQUGou8j2T6KspD2zqy9P6YBXo3t9jaVBvQPFHAYlSjghDT5uoUI6-F6YNSxEeV4xsrm7BKFNIixQlDDgXAoB9y6hxZyti2LS9lySTyEfbcEucsn9gwURZBPSVAZyaxzc6THniZKGffaB+psHlg+LODcAj1H6Kz8Vh3WVx7gkifM9p9T5XjgJfqZsvBDn2geetEoUL8Xvu3lZamSstpHK3GgA)

It's mostly just a copy paste of your code from the sandbox.

**Reply by cbejensen** [reply]
Sorry for the massive delay on this! I don't know why that error is appearing, but I would just [cast it](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgMWAUwDYBMCiUrQDOANCutgGoCGGArmiXAApUwAWpthay0IAwhAB2MNAA94AXzgAzAiDgByKGioBjGAFo2ECAGtNMvooDcAKFCRYiOAHM0U2fKUYIWKoTamL4aPCRcaPw0GABG6npw0nIQCsqqGt5mMACeYGhwAKpCekIQAO5CLOxwALxK2bkFQnBgrGxwmrUYqtxwgXYOxWwAFACUSQD0AFTDZnDDcACyVHoMsnxwarGQQmgihHAqXFShLQB042MTcGwwMGCEAFyDg8tYDFRCWKEQYvvAEIOEg4Gaa4RRFhDHxNAA3ABMmhSAEYAMwyOFwgD8MmALVK3ygakGAEEwGB9jBCGIjkc4AABOpQKgKanrGDdRpwdgZOolGIKVlbXTwCAyFlsDJGKBciCC4CbAFAhaiw4TckU8S0sAtckAA01xNJJ0sfhsDxkVFoGBguKwWBUhEIvFF1DoDFI5stDBti2izkU+zxFqtbtF3l1vmsSENxtNADlaTw+Pb6IwoyAY6Kok5YkpvYnkyBA8dJsshIC4GGTYy0FBCMJbSA4-NyghyXA4EJo1di2gjaWs9XayRG3AqL7XW2S6bnX6ezR4+TJOYFScZLQhBpPjUmOXK0Jqz0kCKQEzJG2IwwgdWmARLn1EP2C0WkPYYHgCBXSA+D2V2txj4C0FhqwAeVJ0n5dtO1NdcKyrWMpwYAA+Ho926Po50mJtb3gFskyfIgPwfbCKx6RRMLQRRkLgW44FadEUk6eBy2fQgyROJsVBgWgoBqHp+ybf80UwLBuBgWDuKbOB-xaexnlgiDN3-QYJPWLBhOY0SeO7RZEPqUoEDfepCOI0jpEGZTUNUxBiPwwhJBEnjx1datZX3LSdK6PTFEHF1rUM8iTLMuS+OwQTfKbZCZw1TUjkGMxxCseBF2XGBV0-NBv1PPh-1S39e0oiRFM2VB+N7eCzAASAozI2nVQEoGAIRbDgAAfdpng7Wrf3VOBap-Qc4BA5EWXFEBZgyQh2JG2IHDYWr6slWoPG4LBOpqDxOpgRRNlaGiYHFEUH1K+kRG6NtqumxrmsNNqsDMK8GxK9DHI-QJq0EERxBgf8ACU0GWKAsH-So8kKbpSEyrAivg0LSvu3SSnKQJggwMIIi4kqSv-M96hy0Rnk2boMpPLKYMIWCzqXC61kWuGWrRCnitRu6puwbpkTbDH2FKkq+lZvgmTevLztaim4H6gHqiZNt1QAEgQUWgfqSR9mliNhAjE0MF2FpALSNAQLURmsG6WDJA60oSdu1GKIAcXWctgDUTQ+eeU6l1XTY1CePJ4FCDIW3wApf2ZM4LmuW5bGAdhaFCfZlhAQYpjtghKxkGBBgAFW1gBlNQarAFPJVGhhBnhABOYuAFYObKwZKMIDBaq0LBJQ1tB-jezQ67WSkgIYbPgFzh3a-rwY8gdsRVTt8PNCeFJK9Y9iaj19EDcx-qACsIFq7pCB6A7GXqUhF6Z+orzbHpD+XkoVun0LUckYgOYAbV37oAF1Sshu7hCLPC-YrR7uARkjNQegUZozZg0R2AlmD1Hxj+MGRMSZNTJoLAOVNya-jpqjc+zNuaim6BzK8ptrz0znhxWiCE+D7D3BnGArA0D7HokQV8rl2Bn31khK8K1wFY35sgmmAd+oFWwJZWBQIipwDbNbGAAAhFIeMhG4F-oQURhMHTE1IOA2C5hb731Rg-PcVC+A0LoQwpRzC97sDfpzcwpVSE1CQPsRxe5zGWXMQecw1kzAjDzHAYINRvY8JaotUINEUgQHYrUC8mxDTyh8UqMQKo1QnE1OqYk5IYr6nruWI0agMgySgngqJgSoEVRSgTP8PMikNhUolGALRjowBqnVFCUQmKTHiiuYQzANwFJADuRywMWThxaFENs+StyVIgJea8Kl7p7n-mUuB25NLsFCicayySIoTCihk6wWSoA5IyKDM8VTSreLgKnIUc0OTOG5AQCAfIBTcnmdtCUUpymOX2L4p4cAAmxHDjKYATyrnnwegWRpEAMCbE8OE7AA4CSqCgAOeA3JCDRjgC0MEmABybG5LvL5wwoolXZGwFmcATrNLMJ4rxoxySlMWq89etU5oqBEA9Elmx8jhwaKC+ZHLYmKmVOAJJkwUlpJOEyoo9Rt7uX2IoUgXoo7yqUHCBVWA5VkQonYzYsqo5wn2FgRQ4V1SRWimPfUHTEpdMlVvHojj9gctPhS+qTUhC0BAN7JFSDqaXTOm6xGfQH4vy5uSxpp1bp2OuWwRiqN9iSsIhq8i1cPJwC+d7GA+Q0DrEouoBoJKOb7BUKqdQaAeiDAADr7AQBCYgkhBi2AVYmrVaBi25NTV8rlJRDixqLerXJZaAB6lbBgKtIkmrYaAQAQCxRi1Qjc6qpoLb2ktZbK2SxHUoUiJhx0qCnTOxpVB0SnUOJ47xdKFoDVogOdsLaMA0WlAHbucAwTACoNesABAX0PEWs6qNn5TpYAec2B5rBVwCpOPExJaBjXivzF-PkoRV4fiQOkSCQg2xIGIg0ppDaBz2DbG6j15YoiznJE+3EH4pGyLxhARDCrUObkUFo8dmHWyhpw9uqg+Hmzus9a0k4T7pGUYcNRmBtHV70Z6UIfYBlmMUWdWR7WvjhMyLkWJujSgGPCH2Fxkicnq6Ec9Yp9IcAAAiKnRPsH-OJyTaGPhCDBDQYAhr9MC34VdTZJqThB0uDcQYDwwREi+IlS4KQAD6AXBiDkc8uX8mhu6EF7v3FQtUsUiGmu3NwHg2CaAfJoAALDIIUprdkoqU1RtTVmADyGnDYfl5rlHGcApYIFqjIYjAAZDsMAFbSza8Rj6wBbBnGNqVfqXXk7FM2HMMJAoaurw5v1CreMcBiDUHQB4-55sPwmzAF+nAfW01IIN4bQkzqrYPRoLbiGdvdf225y6wlUZtj4ZdUqYzMaQOm2gWbcB5tjb+zd1+72HsUxMEAA)

**Comment by NixBiks** [comment]
I'm a bit confused on the state of this. Is this currently possible in v7? And is it "cancelled" in v8 due to [this](https://github.com/react-hook-form/react-hook-form/pull/5941)?

**Comment by Evgenyguzman** [comment]
Hi, any update on this? do we need to wait for v8 and are there any plans for v8 release?

**Comment by BrendanC23** [comment]
What is the current status of this? [This PR](https://github.com/react-hook-form/react-hook-form/pull/5941) looks like it does some of the implementation, but it was closed in June 2022.

**Comment by aspirisen** [comment]
It looks like similar to what I suggested some time ago https://github.com/react-hook-form/react-hook-form/issues/9289

**Comment by aspirisen** [comment]
Hi, I created a PR that demonstrates a Lens approach for building reusable components with TypeScript support https://github.com/react-hook-form/react-hook-form/pull/12284
